odk/src/main/java/org/motechproject/odk/parser/impl/XformParserODK.java

```diff
@@ -0,0 +1,206 @@
+        for (int i = 0; i < nodeList.getLength(); i++) {
+            Node element = nodeList.item(i);
+
+            if (element.hasChildNodes()) {
+
+                if (element.hasAttributes()) {
+                    String localUri = uri + "/" + element.getNodeName();
+                    FormElement formElement = new FormElementBuilder()
+                            .setName(localUri)
+                            .setLabel(localUri)
+                            .setChildren(new ArrayList<FormElement>())
+                            .setType(FieldTypeConstants.REPEAT_GROUP)
+                            .setPartOfRepeatGroup(parent.isRepeatGroup() || parent.isPartOfRepeatGroup())
+                            .createFormElement();
+                    formElementMap.put(formElement.getName(), formElement);
+                    group.add(formElement);
+                    recursivelyAddGroup(formElementMap, element.getChildNodes(), uri + "/" + element.getNodeName(), formElement.getChildren(), formElement);
+
+                } else {
+                    recursivelyAddFormElements(formElementMap, element.getChildNodes(), uri + "/" + element.getNodeName());
+                }
+
+            } else if (element.getNodeType() == Node.ELEMENT_NODE) {
+                String localUri = uri + "/" + element.getNodeName();
+                FormElement formElement = new FormElementBuilder()
+                        .setName(localUri)
+                        .setLabel(localUri)
+                        .setPartOfRepeatGroup(parent.isRepeatGroup() || parent.isPartOfRepeatGroup())
+                        .createFormElement();
+                group.add(formElement);
+                formElementMap.put(formElement.getName(), formElement);
+            }
+        }
+
+    }
+
+    protected boolean hasChildElements(Node element) {
+        NodeList children = element.getChildNodes();
+        for (int i = 0; i < children.getLength(); i++) {
+            if (children.item(i).getNodeType() == Node.ELEMENT_NODE) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private List<FormElement> createFormElementListFromMap(Map<String, FormElement> formElementMap) {
+        List<FormElement> formElements = new ArrayList<>();
+        for (FormElement formElement : formElementMap.values()) {
+            if (formElement.getType().equals(FieldTypeConstants.REPEAT_GROUP) || !formElement.isPartOfRepeatGroup()) {
+                formElements.add(formElement);
+            }
+        }
+        return formElements;
+    }
+
+    private void removeFormNameFromLabel(Map<String, FormElement> formElementMap) {
+        for (String key : formElementMap.keySet()) {
+            FormElement element = formElementMap.get(key);
+            String label = element.getLabel();
+            element.setLabel(label.substring(label.substring(1).indexOf('/') + 1));
```

Comment:
After reviewing the whole find I came to the conclussion that this substring parsing belongs in some sort of FormUtils (or something along those lines) class.